### ISSUE TITLE
Improve interoperability by moving Request and Response to another crate

### ISSUE THREAD

**Issue by s-panferov** [iron/iron#295]
Hello @reem. I have found myself in creating my own Request and Response traits in Rustless because I want to be able to work with other (potential) backend libraries. 

What if we move these items into a separate crate and unify this layer in nickel, iron, restful, rustless and any future libraries?

cc @cburgdorf, @Ryman, @Ogeon

**Comment by Ogeon** [comment]
I'm interested. It's obvious that we have slightly different approaches to the problem, but our libraries exposes mostly similar request handling. Having a common API would make it easier to use the same additions on top of different base libraries. It fits well with the Rustful philosophy :)

The big question is "How?"

**Comment by s-panferov** [comment]
I think that we can pass it through an RFC process. Maybe I could write the RFC if you forgive my poor English. But at first I want to hear that all are interested.

**Comment by sunng87** [comment]
In #261, I have a proposal that make iron decoupled with hyper or any other backend. I think it shares the same idea of this one. Iron can be a standard request/response and middleware framework. The will be lower level adapters for hyper, and higher level frameworks for url routing, session/cookies, etc.

**Comment by Ogeon** [comment]
The problem I can see right now is that we will have to unify the way we handle data. It may even require an adapter layer. Rustful, for example, uses Hyper as a base and extends the request with more context and the response with a filtering plugin stack. I would rather not drop them from the library and I'm sure each library has its own complicating factor. Nickel, for example, is still using its own version of rust-http (right?) and Iron seems to be built to create the response inside the handler and return it.

**Comment by s-panferov** [comment]
@sunng87 

> Iron can be a standard request/response and middleware framework

I don't think that we must and *need* to peek a standard framework, because all people has their own opinion. Time will put everything in its place. But we are able to pick some small standard parts now. Let's begin with that.

**Comment by sunng87** [comment]
@s-panferov Yes, there's no such "standard" thing. The idea is to make iron decoupled from underlying network layer. That is, do not use anything "hyper" specific in the request and response struct. 

Currently there's no need to persuade other framework like nickle to adopt. Just to make iron API more beautiful.

**Comment by sunng87** [comment]
https://github.com/conduit-rust/conduit 

Conduit is aimed to be such a library, not sure if iron has plan to adopt.

**Comment by Ryman** [comment]
Very interested in this, it's crossed my mind a few times, and I agree that 'how' is the question. 

Rusts ownership model makes this tricky as @Ogeon has pointed out, where `iron` creates Responses in the handler while others pass them down into the handlers. At least we could start with a common base of traits based on hypers request and its `Streaming` and `Fresh` responses?

@Ogeon re: rust-http: Yeah, still on a fork, but will hopefully be using hyper in the next week or so. 

@sunng87 things have probably changed a lot since then, but @reem did leave feedback on conduit's issue tracker [back in November](https://github.com/conduit-rust/conduit/issues/4) where he considers a potential common base.

cc @SimonPersson

**Comment by Ogeon** [comment]
I almost forgot about Conduit for a moment. It can surely function as a minimal common set of features. I mean, it's all just HTTP in the end. The trick is to identify this set and make sure the interface is flexible enough to avoid turning everything into Hyper+overhead.

**Comment by reem** [comment]
I agree we want this long term, but I honestly don't think we really know enough about what Rust web development will look like for us to make these decisions right now.

**Comment by reem** [comment]
For what it's worth I own the `http` name on crates.io - that seems to me to be a fine place to aggregate standards for web development in rust.

**Comment by lucab** [comment]
Any progress on this? I'm approaching the same issue from the other end (fastcgi / uwsgi / etc connectors) and I'm not sure which traits should be provided to applications (for headers/env/body) and what should expect back.

This is an open question for [uwsgi-rust](https://github.com/unbit/uwsgi-rust), which is currently abusing strings everywhere.

**Comment by Ogeon** [comment]
No, not as far as I know. You could possibly reuse pieces of some general HTTP library, in your case, but it would probably come with a lot more than what you need.
